src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt

```diff
@@ -0,0 +1,66 @@
+package uk.co.llobile.kotlin.hours.types
+
+sealed class FunctionalOption<T>: Functor<T>, Applicative<T>, Monad<T> {
+
+    abstract fun isEmpty(): Boolean
+    abstract fun get(): T
+
+    override fun <R> map(fn: (T) -> R): FunctionalOption<R> {
+        return if (isEmpty()) None() else {
+            Some(fn(get()))
+        }
+    }
+    override fun <R> apply(fn: Applicative<(T) -> R>): FunctionalOption<R> {
+        return if (isEmpty()) None() else {
+            fn.map { it(get()) } as FunctionalOption<R>
+        }
+    }
+    override fun <R> lift(fn: (T) -> Monad<R>): FunctionalOption<R> {
+        return if(isEmpty()) None() else {
+            fn(get()) as FunctionalOption<R>
+        }
+    }
+}
+
+class Some<T>(val value: T): FunctionalOption<T>(){
+    override fun isEmpty() = false
+    override fun get() = value
+
+    override fun toString(): String {
+        return String.format("Some(%s)", value)
+    }
+
+}
+
+class None<T>() : FunctionalOption<T>() {
```

Comment:
I don't think this should be a class. It should be a not generic **object**. if you set the `FunctionalOption` as covariant you can set the `None` as `FunctionalOption<Nothing>`

Reply:
Done